README.md
# Neodim Server

Neodim Server is an HTTP-server
that provides natural language text generation
using a selected language model.

---

Main website: https://github.com/alkatrazstudio/neodim-server

Current stable version: **v0.13** (August 6, 2023) • [CHANGELOG](CHANGELOG.md)

Repository branches:
* [master](https://github.com/alkatrazstudio/neodim-server/tree/master) - current stable version
* [dev](https://github.com/alkatrazstudio/neodim-server/tree/dev) - ongoing development

---

**NOTE:** before reading any further you must have at least a general grasp of the following topics:
machine learning, text generation, language model, HTTP server, command line interface.
Also, running this on Linux is recommended.
If you have any other OS you need to install Python, Git, Bash, sed and md5sum.


## Contents

- [How it works](#how-it-works)
- [Security and production use](#security-and-production-use)
- [Requirements](#requirements)
- [Quick Start](#quick-start)
- [Supported models](#supported-models)
  - [Notable models](#notable-models)
- [Downloading the language models](#downloading-the-language-models)
  - [Auto-download](#auto-download)
  - [Download via Git](#download-via-git)
  - [Manual download](#manual-download)
- [CLI](#cli)
- [Prompt and preamble](#prompt-and-preamble)
  - [Recommendations](#recommendations)
  - [Example](#example)
- [API request](#api-request)
- [API response](#api-response)
- [Squeezing the VRAM](#squeezing-the-vram)
  - [Running in console mode](#running-in-console-mode)
  - [Distributing model layers on CPU](#distributing-model-layers-on-cpu)
  - [Lowering the precision](#lowering-the-precision)
- [Third-party libraries](#third-party-libraries)
- [License](#license)
- [Client applications](#client-applications)


## How it works

Neodim Server uses [Transformers](https://huggingface.co/docs/transformers/index)
to provide the actual inference (text generation).
It can use some [language models from Hugging Face](https://huggingface.co/models?library=pytorch&pipeline_tag=text-generation&sort=downloads)
(it will download them automatically if needed)
or locally available models (pre-downloaded from Hugging Face or from anywhere else).
See below for a list of supported models.

After the model is loaded, the web server is started.
You can pass the text generation parameters to it,
and the server will use the model to generate some text and then the server will return this text (or an error).

Neodim Server is a stateless server.
It will not save anything between requests.
All required information will be passed with the request.


## Security and production use

Neodim Server as-is should only be used locally or internally.
It's not designed to be exposed over the Internet or used by multiple clients simultaneously.
For example, it uses the built-in Python HTTP web-server
(which is usually not tested for any security vulnerabilities),
can only handle one request at a time
and can easily be crashed or hanged (e.g. if you pass incorrect `Content-Length` header).

Neodim Server also lacks a lot of sanity checks and is not foolproof to use.


## Requirements

* CUDA-enabled GPU (e.g. any NVIDIA GTX/RTX), supporting at least CUDA v11.8.
* Appropriate amount of free VRAM (e.g., at least 5GB of free VRAM for `GPT-J-6B` model)
* Appropriate amount or free RAM (e.g., at least 15GB of free RAM for `GPT-J-6B` model)
* Python 3.9+ (check with `python3 --version`)
* Preferably Linux (Windows may work, but the compatibility is not tested and not maintained)

You may also need to update your GPU drivers.
On Ubuntu you can do it with `sudo ubuntu-drivers install`.
To download the repo and (optionally) models, you will need Git.
On Ubuntu you can install it with `sudo apt install git`.
To run the server the recommended way (via `start.sh`, see below)
you will need Python Virtual Environment tools.
On Ubuntu you can install them with `sudo apt install python3-venv`.


## Quick Start

```sh
# WARNING: this may download up to 2GB of data
git clone https://github.com/alkatrazstudio/neodim-server
neodim-server/start.sh --model=EleutherAI/gpt-neo-125M
```

This will start a server with the smallest language model:
[GPT-Neo-125M](https://huggingface.co/EleutherAI/gpt-neo-125M).
This model is not suitable for any adequate text generation,
but it will allow you to test if Neodim Server works at all on your machine.

Open `playground.html` in your browser and play with parameters, etc.
All CLI options and parameters are explained below.

Here is another example:
```sh
# WARNING: this may download up to 13GB of data
neodim-server/start.sh --model=EleutherAI/gpt-j-6B --model-revision=float16 --listen-address=0.0.0.0 --layers=14 --precision=float4
```
It will use the official [GPT-J-6B](https://huggingface.co/EleutherAI/gpt-j-6B) model
by [EleutherAI](https://www.eleuther.ai)
from the [`float16` branch](https://huggingface.co/EleutherAI/gpt-j-6B/tree/float16).
Then it will be dynamically converted to 4-bit float to reduce VRAM usage.
The server will also listen on all network interfaces (`--listen-address=0.0.0.0`)
which allows any external client to connect to the server.
It will also only load half of the model to the GPU, and the rest will be on the CPU.

If you have `curl` installed you can try to manually send an HTTP request:
```sh
curl -Ss --data '{
  "prompt": "I walk across the dead train yard, remembering who we are.",
  "generated_tokens_count": 32,
  "max_total_tokens": 128,
  "temperature": 0.6,
  "tfs": 0.95,
  "truncate_prompt_until": [" "],
  "repetition_penalty": 3
}' http://127.0.0.1:8787/generate | jq -r '.sequences[].generated_text'
```
(`jq` is used here just to display the result in a human-readable form,
but you can remove it to view the raw JSON response)


## Supported models

* [GPT-Neo](https://huggingface.co/models?other=gpt_neo),
  [GPT-J](https://huggingface.co/models?other=gptj),
  [GPT-NeoX](https://huggingface.co/models?other=gpt_neox),
  [GPT 2](https://huggingface.co/models?other=gpt2),
  [OPT](https://huggingface.co/models?other=opt),
  [CodeGen](https://huggingface.co/models?other=codegen),
  [BLOOM](https://huggingface.co/models?other=bloom),
  [LLaMA](https://huggingface.co/models?other=llama)
  \- fully supported, except some corner cases.
* [XGLM](https://huggingface.co/models?other=xglm) - only some `fairseq` models are supported.
  You may try your luck with
  [KoboldAI fairseq-dense-* models](https://huggingface.co/KoboldAI).

However, each model may have some quirks which are out of the scope of this documentation.
For example, XGLM and OPT models have trouble with whitespace and newlines.


### Notable models

* [EleutherAI models](https://huggingface.co/EleutherAI) -
  original `GPT-Neo`, `GPT-J` and `GPT-NeoX` models,
  not specifically trained for any task
  (jack of all trades, master of none)

* [KoboldAI models](https://huggingface.co/KoboldAI) -
  models that are used in [KoboldAI](https://github.com/KoboldAI/KoboldAI-Client),
  trained for various tasks (mostly adventure games and novel writing)

* [sberbank-ai/rugpt3large_based_on_gpt2](https://huggingface.co/sberbank-ai/rugpt3large_based_on_gpt2) -
  at the time of writing this document,
  this was the best Transformers-compatible model that can generate Russian text


## Downloading the language models

There are at least three ways to get a language model on your PC.


### Auto-download

Let Neodim Server download the model from HuggingFace.
In this case you need to pass the model's name (e.g. `EleutherAI/gpt-neo-2.7B`)
to the [--model](#model-string-required) parameter.


### Download via Git

Specifying a model's name may be convenient, but it has some drawbacks:
* You need an internet connection even if the model was previously loaded
* It's hard to move the downloaded model to a different place or make a backup.

So, you may want to download the model manually first.
You can do it with Git, for example:
```sh
# first enable LFS if it's not enabled already
git lfs install

# then clone the model's repository
git clone https://huggingface.co/EleutherAI/gpt-neo-2.7B

# or if you want to clone a specific branch
git clone -b float16 https://huggingface.co/EleutherAI/gpt-j-6B
```

After that you'll be able to specify the cloned directory path in the `model` parameter.

**NOTES:**

* The whole Git repository will take twice the space that is really needed to run the model.
  The `.git` subfolder will contain the entire copy of the language model.
  If you don't plan on frequently updating the model you can remove the `.git` subfolder to save some disk space.

* The repository may contain some other files that is not needed for Neodim Server.
  For example, a language model for Rust or something like that.
  The list of required files depends on a model,
  but the only big file(s) that is used by Neodim Server are `pytorch_model*.bin` or `model-*.safetensors`.
  If there are some other multi-gigabyte files (e.g. `rust_model.ot` or `flax_model.msgpack`),
  you can safely remove them.


### Manual download

Downloading models with Git is easy, but as discussed in the previous section,
you may need to do some cleanup afterwards.

It's also possible to manually download all needed files from HuggingFace.

1. Go to the model's page, e.g. https://huggingface.co/EleutherAI/gpt-j-6B.
2. Go to "Files and versions" tab.
3. Choose the required branch from the dropdown, e.g. `float16`.
4. Download all needed files by clicking the "down arrow" icon.
5. Put all these files in the same folder.

Like it was said in the previous section, the only multi-gigabyte file(s) you need are `pytorch_model*.bin` or `model-*.safetensors`.
All other huge files are not needed.
However, all other small files (e.g. `*.json` and `*.txt`) are usually needed to run a language model.


## CLI

To start the server run the `start.sh` script.
You can pass some parameters to it in the following forms: `--<param-name>=val` or `--<param-name> val`.
For example, `--model=EleutherAI/gpt-neo-1.3B` or `--model EleutherAI/gpt-neo-1.3B`.

Pass `--help` or `-h` to show a short help.
Pass `--version` to show the version of your Neodim Server.

Below is the list of all supported parameters.


### `model`: string (required)

You can specify either a path to the directory of your model,
or its name on Hugging Face (e.g. `EleutherAI/gpt-neo-2.7B`).

The directory path must either be absolute (e.g. `/path/to/model`)
or start with a dot (e.g. `./some/path` or `../some/path`).

Read more on how to [download the language models](#downloading-the-language-models).

### `model-revision`: string (optional, default=\<default branch for the repo\>)

If you specified a Hugging Face model name in the `model` parameter
then `model-revision` will let you specify a specific branch.

By default, it will download the main branch.

### `cache-dir`: string (optional, default=\<transformer's default\>)

A directory for storing downloaded models.

By default, it's chosen by `transformers` framework.

### `listen-address`: string (optional, default="127.0.0.1")

Listen on this IP address.

### `listen-port`: int (optional, default=8787)

Listen on this port.

### `layers`: (int|"a")[] (optional, default="a")

Distribute model's layers on GPUs.

It's a comma-separated list of integers.
Each integer represents how many layers should be put on a GPU with that index.
All unspecified layers will be put on CPU.

You can also use a letter `a` (stands for "**a**vailable")
for one GPU to put all unspecified layers on it.
For example, `--layers=a` will put all model's layers on the first GPU.
`--layers=2,a,3` will put 2 layers on the first GPU, 3 layers on the third GPU
and the rest of the layers on the second GPU.

GPUs and their indexes are shown at server startup.
The number of layers is shown after the model is loaded,
but you also can find this number in `config.json` of the model
(see `n_layer`, `num_layers` or `num_hidden_layers` fields).

For example:
```sh
./start.sh --model=KoboldAI/GPT-J-6B-Skein --layers=0,2
```
This will put no layers on the first GPU, 2 layers on the second GPU and the rest of the layers on the CPU.

The more layers you put on GPUs the faster the model will work
and the more VRAM it will use.
You can also put all layers on CPU (`--layers=0`)
or put all layers on a single GPU (e.g. `--layers=a`, `--layers=0,a` etc.).

### `precision`: original|float32|float16|float4|int8|gptq2|gptq4|gptq8 (optional, default=original)

Neodim Server can load 16-bit and 32-bit models,
but using this parameter the model can be load in memory with a specified precision.
This allows to seek balance between VRAM consumption and the quality of the generated text.

* `float16` - 16-bit floating precision, a.k.a. "half-precision".
* `float32` - 32-bit floating precision, a.k.a. "full-precision".
  Compared to `float16`, loading the model in `float32` may improve the quality of the generated text,
  but the model will consume nearly twice as much VRAM.
* `int8` - 8-bit integer precision.
  Compared to `float16`, loading the model in `int8` may lower the quality of the generated text,
  but the model will consume around 1.5 times less VRAM.
* `float4` - 4-bit float precision.
  Compared to `int8`, loading the model in `float4` may lower the quality of the generated text,
  but the model will consume around 1.5 times less VRAM.
* `gptq2`, `gptq4`, `gptq8` - [read below](#gptq)
* `original` - use the original precision of the model.
  This is the default.

It only makes sense to downgrade the model's precision.
For example, if you load 16-bit model in 32-bit precision mode
then it will consume twice as much VRAM, but it won't improve the quality of the generated text.

#### **INT8**

1. You need to install CUDA Toolkit.
   There are several ways to do it:
   - On Ubuntu: `sudo apt install nvidia-cuda-toolkit`
   - The official NVIDIA page: https://developer.nvidia.com/cuda-downloads
   - [Conda](https://docs.conda.io/): `conda install -c nvidia/label/cuda-<cuda version> cuda`
2. CPU-layers won't be converted to 8-bit.
3. You can load 8-bit models directly, but you still need to specify `--precision=int8`.
   Also, in this case all layers must be on GPU, e.g. `--layers=a`.
4. Some models may fail to load completely.

#### **FLOAT4**

The above notes apply to `float4` precision as well.

Also, when using `--precision=float4` the following additional CLI parameters are available:
[--quantization-type](#quantization-type-fp4nf4-optional-defaultnf4)
and [--double-quantization](#double-quantization-truefalse-optional-defaulttrue).

#### **GPTQ**

Neodim Server supports 2-bit, 4-bit and 8-bit models quantized with [GPTQ](https://arxiv.org/abs/2210.17323).
4-bit models consume roughly 1.5-2 times less VRAM than 8-bit models,
and only have slightly lesser quality (the bigger the model - the less the quality loss).
On HuggingFace Hub such models will have either "GPTQ" and/or "4bit" and "128g/64g/32g" in their name, but not "GGML" (that's a different quantization technique, not compatible with Neodim Server).

Example 4-bit GPTQ model: https://huggingface.co/TheBloke/vicuna-13B-1.1-GPTQ-4bit-128g.

To load GPTQ models you may want to specify additional parameters to CLI:
[--model-basename](#model-basename-string-optional),
[--group-size](#group-size-int-optional),
[--true-sequential](#true-sequential-truefalse-optional-defaulttrue)
and [--safetensors](#safetensors-truefalse-optional-defaulttrue).
The last three already have popular defaults, so most of the time you only need to specify `--model-basename`.
Also, if there's a `quantize_config.json` file inside the model's folder,
then you don't need to specify `--group-size` and `--true-sequential`.
If the GPTQ model has a proper filename, then `--model-basename` is also not needed.
There are also some additional parameters that can influence the model behavior:
[--fused-attention](#fused-attention-truefalse-optional-defaultfalse)
and [--fused-mlp](#fused-mlp-truefalse-optional-defaultfalse).

Current limitations and gotchas:

* Some models may not load, it depends on how they were quantized.
* `--precision=gptq8` is different from `--precision=int8`,
  and `--precision=gptq4` is different from `--precision=float4` - GPTQ uses a different quantization algorithms, so it's not possible to load a GPTQ model with `int8` and `float4` precisions, and vice versa.
* `--precision=int8` and `--precision=float4` can quantize the model on fly, i.e. you can load float16 model in INT8 precision. GPTQ models, on the other hand, need to be quantized beforehand.
* You will need `tokenizer.json` for GPTQ models, otherwise they will be very slow.

Example that shows how to load the abovementioned model:
```sh
./start.sh --model=/opt/models/vicuna-13B-1.1-GPTQ-4bit-128g --precision=gptq4 --model-basename=vicuna-13B-1.1-GPTQ-4bit-128g.latest
```

You can also load GPTQ models directly from the HuggingFace Hub:
```sh
./start.sh --model Neko-Institute-of-Science/LLaMA-7B-4bit-128g --precision=gptq4 --model-basename=llama-7b-4bit-128g
```

Loading a GPTQ model that has the `quantize_config.json` file and a proper filename:
```sh
./start.sh --model /opt/models/Llama-2-13B-chat-GPTQ
```

### `model-basename`: string (optional)

Specify a model basename, i.e. the filename without directory and extenstion.
This setting only applied for [GPTQ](#gptq) models, i.e. if `--precision=gptq*`.

For example, to load the model `/some/dir/wizardLM-7B-GPTQ.safetensors`,
use this: `--model=/some/dir --model-basename=wizardLM-7B-GPTQ --safetensors=true`.

You can omit this parameter if the GPTQ model is named appropriately,
e.g. `gptq_model-4bit-32g.safetensors`.

### `group-size`: int (optional)

The group size parameter for a [GPTQ](#gptq) model.
This value is usually mentioned in the description of the model.
Sometimes it's in the filename - a number with a `g` suffix,
e.g. `koala-13B-GPTQ-4bit-128g` has group size of `128`.

This parameter is ignored if there's a `quantize_config.json` in the model's folder.

### `true-sequential`: true|false (optional, default=true)

The "true sequential" flag for a [GPTQ](#gptq) model.
This value is usually mentioned in the description of the model.

This parameter is ignored if there's a `quantize_config.json` in the model's folder.

### `safetensors`: true|false (optional, default=true)

Whether to load [GPTQ](#gptq) model in `safetensors` format.
If `false`, then the `*.bin` or `*.pt` file will be loaded.

### `fused-attention`: true|false (optional, default=false)

Inject fused attention.
Only applicable to [GPTQ](#gptq) models, but may not be incompatible with some of them.
May increase the inference speed at the cost of higher VRAM usage.

### `fused-mlp`: true|false (optional, default=false)

Inject fused MLP.
Only applicable to [GPTQ](#gptq) models, but may not be incompatible with some of them.
May increase the inference speed at the cost of higher VRAM usage.

### `quantization-type`: fp4|nf4 (optional, default=nf4)

Use [4-bit floating point](https://huggingface.co/blog/4bit-transformers-bitsandbytes#fp4-precision-in-a-few-words) (`fp4`)
or [4-bit NormalFloat](https://arxiv.org/abs/2305.14314) (`nf4`)
when performing the quantization.
This mayt affect the quality of the generated text.
Only applicable to `float4` precision.

### `double-quantization`: true|false (optional, default=true)

Perform double quantization.
Disabling it will reduce the VRAM consumption by approximately 10%,
but may lower the quality of the generated text.


## Prompt and preamble

With each request you can pass two chunks of text to the model.
The first chunk is "preamble" and the second chunk is "prompt".

The final input text will be the result of concatenation of preamble and prompt
(without any spaces).
However, prompt may be truncated from its beginning
so that the final input text can fit inside the total allowed number of tokens.
The preamble is never truncated.

Either preamble or prompt can be empty, but not both at the same time.

Both preamble and prompt may be modified before they are passed to the language model.
It can happen if they contain special tokens that are handled differently by the model.
See the response parameters
[used_prompt](#used_prompt-string) and [preamble](#preamble-string)
to see what was actually passed to the model.


### Recommendations

A lot of models give incorrect results if the input text has spaces in particular places.
Below are the list of recommendations to reduce the number of inference errors.

* Strip all leading and trailing spaces from the input text (`prompt + preamble`).
  It's safe to leave leading and trailing newlines, though.
* Squish multiple consequent spaces into one.
* Do not put any spaces right after a newline.
* For newlines only use `\n` and not `\r`.

Check each model output independently to see what happens when you put spaces in various places.


### Example

Let's assume we have this preamble:
```
This is a story about Alice.


```
(notice the two line breaks)

And the prompt:

CHANGELOG.md
# Neodim Server - CHANGELOG


## v0.13 (August 6, 2023)

- Changed: GPTQ settings file (`quantize_config.json`) is now prioritized over the command line options
- Added: [words_blacklist_at_start](README.md#words_blacklist_at_start-string-optional) parameter


## v0.12 (June 11, 2023)

- Added: support for [GPTQ](README.md#gptq) models.
- Added: support for `float4` [precision](README.md#precision-originalfloat32float16float4int8gptq2gptq4gptq8-optional-defaultfloat16)
- Added: [can_stop_early](README.md#can_stop_early-bool-optional-defaultfalse) parameter
- Changed: switching to CUDA 11.8
- Changed: all layers are distributed on the first GPU by default
  (if [--layers](README.md#layers-inta-optional-defaulta) is not specified)
- Improved: `int8` precision is now supported on all GPUs supported by Neodim Server
- Improved: blacklist processing


## v0.11 (April 23, 2023)

- Added: [no_repeat_ngram_size](README.md#no_repeat_ngram_size-int-optional) parameter
- Added: [words_blacklist](README.md#words_blacklist-string-optional) parameter
- Added: [words_whitelist](README.md#words_whitelist-string-optional) parameter
- Added: support for [LLaMA](https://huggingface.co/models?other=llama) models
- Added: support for safetensors
- Improved: 8-bit models can now be loaded directly
- Fixed: the playground misses some input parameters
- Fixed: multiple encoding/decoding problems
- Changed: the server now tries to generate at least one token


## v0.10 (March 4, 2023)

- Fixed: specifying `top_k` gives a type error
- Added: contrastive search
  (see [penalty_alpha](README.md#penalty_alpha-float-optional) parameter)


## v0.9 (February 19, 2023)

- Fixed: wrong truncation of the inference result
- Changed: special tokens (e.g. `<|endoftext|>` or `</s>`) are now always removed
  from the input and the output
- Added: server version validation
  ([required_server_version](README.md#required_server_version-string-optional) request parameter)
- Added: full output from the model is returned in the
  [output_text](README.md#sequencesoutput_text-string) response field
- Added: actually used preamble is returned in the
  [preamble](README.md#preamble-string) response field
- Improved: silenced messages: "the specified maximum sequence length" and "Welcome to bitsandbytes"


## v0.8 (December 18, 2022)

- Changed: using CUDA 11.7
- Changed: `stop_strings` and `truncate_prompt_until` are not sorted anymore
- Added: support for
  [GPT-NeoX](https://huggingface.co/models?other=gpt_neox),
  [CodeGen](https://huggingface.co/models?other=codegen) and
  [BLOOM](https://huggingface.co/models?other=bloom) models
- Added: ability to specify regular expressions as stop strings (see
  [stop_strings_type](README.md#stop_strings_type-enumstringregex-optionaldefaultstring),
  [stop_strings_required_matches_count](README.md#stop_strings_required_matches_count-int-optional-default1),
  request parameters and
  [stop_string_match](README.md#sequencesstop_string_match-string)
  response parameter)
- Improved: layers distribution is now supported for 8-bit precision
  (i.e. `layers` can be set to any supported value when `precision=int8`)
- Improved: more readable display of free VRAM and layers distribution
- Improved: repetition penalty can now be specified in
  [warpers_order](README.md#warpers_order-string-optional)


## v0.7 (September 18, 2022)

- Added: ability to load the model in 32-bit and 8-bit precisions
  ([precision](README.md#precision-originalfloat32float16int8-optional-defaultfloat16) CLI param)
- Fixed: fast tokenizer is not working for OPT models


## v0.6 (July 16, 2022)

- Added: typical sampling (`typical` request param)
- Added: top-a sampling (`top-a` request param)
- Added: the order of filters/sampling/warpers can now be set (`warpers_order` request param)
- Added: the playground settings can now be reset to their defaults


## v0.5 (July 2, 2022)

- Changed: using CUDA 11.6 (to support modern GPUs)
- Improved: loading models is now 4 times faster
- Improved: GPT2 models now support layers distribution
- Added: support for OPT models
- Added: use `a` symbol to put all unspecified layers on a GPU


## v0.4 (June 26, 2022)

- Fixed: wrong device detection when moving the entire model to GPU


## v0.3 (April 10, 2022)

- Added: `--version` flag to show the current version


## v0.2 (April 2, 2022)

- Fixed: avoid the crash with penalty slope when penalty range = 1


## v0.1 (March 20, 2022)

- Initial release


src/main.py
# SPDX-License-Identifier: AGPL-3.0-only
# 🄯 2022, Alexey Parfenov <zxed@alkatrazstudio.net>

import argparse
import time
from argparse import Namespace

import torch
from transformers import PreTrainedModel, PreTrainedTokenizer

import ai
import dev_map
import server
import tools
from ai import GeneratedOutput, ModelLoadOptions, ModelPathOptions, ModelPrecision, QuantizationType
from server import Callback, ServerRequestData


def get_request_callback(model: PreTrainedModel, tokenizer: PreTrainedTokenizer, gpu_device: int | None) -> Callback:
    def request_callback(request: ServerRequestData) -> GeneratedOutput:
        tools.cleanup()

        t_start = time.time()

        gen_out = ai.generate(
            model=model,
            tokenizer=tokenizer,
            gpu_device=gpu_device,
            request=request
        )

        t_elapsed = round(time.time() - t_start)

        free_mems = [
            f"GPU_{gpu_index}={tools.format_gb(gpu.memory_free_end)}"
            for gpu_index, gpu in enumerate(gen_out.gpus)
        ]
        free_mems_str = ", ".join(free_mems)

        print(f"Generated {gen_out.generated_tokens_count} tokens in {t_elapsed}s " +
              f"with {gen_out.used_input_tokens_count} input tokens. Free VRAM: [{free_mems_str}]")

        tools.cleanup()
        return gen_out

    return request_callback


def print_gpu_info() -> None:
    title = server.name_and_version()
    print("=" * len(title))
    print(title)
    print("=" * len(title))
    print()

    print(f"Using CUDA v{torch.version.cuda}")
    print()

    gpu_devices_count = torch.cuda.device_count()
    if not gpu_devices_count:
        print("No CUDA GPUs found")
    else:
        print("Found GPUs:")
        for gpu_index in range(gpu_devices_count):
            mem_free, mem_total = torch.cuda.mem_get_info(gpu_index)
            gpu_name = torch.cuda.get_device_properties(gpu_index).name
            mem_total_str = tools.format_gb(mem_total)
            mem_free_str = tools.format_gb(mem_free)
            print(f"{gpu_index}. {gpu_name} ({mem_total_str} total, {mem_free_str} free)")
    print()


def load_model(args: Namespace) -> tuple[PreTrainedModel, PreTrainedTokenizer, int | None]:
    path_options = ModelPathOptions(
        name_or_dir=args.model,
        revision=args.model_revision,
        cache_dir=args.cache_dir,
        basename=args.model_basename
    )

    if not path_options.revision:
        print(f"Loading the model: {path_options.name_or_dir}")
    else:
        print(f"Loading the model: {path_options.name_or_dir} ({path_options.revision})")

    t_start = time.time()

    config = ai.load_config(path_options)
    model_type = tools.model_type(config)

    layers_count = tools.num_layers(config)
    layers = dev_map.parse_layers(args.layer_strs, layers_count)
    layers_str = ", ".join(f"GPU_{device_index}={layer}" for device_index, layer in enumerate(layers) if layer > 0)
    gpu_layers_count = sum(layers)
    cpu_layers_count = layers_count - gpu_layers_count
    if cpu_layers_count > 0:
        layers_str = layers_str + f", CPU={cpu_layers_count}"
    print(f"Total layers: {layers_count} ({layers_str})")
    gpus_count = len([x for x in layers if x])

    device_map = dev_map.build(model_type, layers_count, layers)
    gpu_device = dev_map.find_first_gpu_device(device_map)

    load_options = ModelLoadOptions(
        precision=args.precision,
        group_size=args.group_size,
        true_sequential=args.true_sequential,
        use_safetensors=args.safetensors,
        device_map=device_map,
        gpu_device=gpu_device,
        inject_fused_attention=args.fused_attention,
        inject_fused_mlp=args.fused_mlp,
        quantization_type=args.quantization_type,
        double_quantization=args.double_quantization
    )

    model, tokenizer = ai.load_model(
        path_options=path_options,
        load_options=load_options
    )

    t_elapsed = round(time.time() - t_start)
    print(f"Model {model.__class__.__name__} ({model_type.name}) loaded in {t_elapsed}s")
    if device_map is not None:
        print(f"Distributed {gpu_layers_count} layer(s) to {gpus_count} GPU(s), and {cpu_layers_count} layers to CPU")
    else:
        print("Moving the entire model to CPU")
        model = ai.move_to_cpu(model)

    print()

    return model, tokenizer, gpu_device


def print_free_ram():
    gpu_devices_count = torch.cuda.device_count()
    free_mems = []
    for gpu_index in range(gpu_devices_count):
        memory_free, _ = torch.cuda.mem_get_info(gpu_index)
        free_mems.append(memory_free)
    free_mem_strs = [tools.format_gb(m) for m in free_mems]
    free_mem_str = ", ".join(free_mem_strs)
    print(f"Free VRAM: {free_mem_str}")
    print()


def run_ai_server(
    model: PreTrainedModel,
    tokenizer: PreTrainedTokenizer,
    gpu_device: int | None = None,
    ip: str = "127.0.0.1",
    port: int = 8787
) -> None:
    callback = get_request_callback(model, tokenizer, gpu_device)
    print("Server is running")
    print(f"Endpoint for text generation: http://{ip}:{port}{server.ENDPOINT_PATH}")
    print("Press CTRL+C to stop")
    print()
    print("------")
    print()
    server.run(ip, port, callback)

    print()
    print("------")
    print()
    print("Neodim Server is terminated.")


def parse_args() -> Namespace:
    parser = argparse.ArgumentParser(
        prog="start.sh",
        description=f"{server.name_and_version()} - natural language model AI via HTTP",
        epilog="License: AGPLv3\n"
               "More info and help: https://github.com/alkatrazstudio/neodim-server",
        formatter_class=argparse.RawDescriptionHelpFormatter,
        allow_abbrev=False
    )
    parser.add_argument("--model",
                        help="Path to the model or its name on Hugging Face "
                             "(paths must either be absolute or start with a dot)")
    parser.add_argument("--model-revision",
                        help="Model revision/branch on Hugging Face")
    parser.add_argument("--cache-dir",
                        help="Cache directory for downloading Hugging Face models")
    parser.add_argument("--listen-address",
                        help="Listen on this address (pass 0.0.0.0 to listen on all addresses). "
                             "Default: 127.0.0.1",
                        default="127.0.0.1")
    parser.add_argument("--listen-port",
                        help="Listen on this port. "
                             "Default: 8787",
                        default="8787")
    parser.add_argument("--layers",
                        help="Distribute model's layers between GPUs (pass 0 to not use GPUs). "
                             f"Default: {dev_map.AVAILABLE_LAYERS_CHAR}",
                        default=dev_map.AVAILABLE_LAYERS_CHAR)
    parser.add_argument("--precision",
                        help=f"Load the model in this precision ({', '.join([x.value for x in ModelPrecision])}). "
                             f"Default: {ModelPrecision.ORIGINAL.value}",
                        default=ModelPrecision.ORIGINAL)
    parser.add_argument("--model-basename",
                        help="Filename without extension of the model "
                             "Currently only supported for GPTQ precisions.")
    parser.add_argument("--group-size",
                        help="Group size for GPTQ. "
                             "Default: 128",
                        default="128")
    parser.add_argument("--true-sequential",
                        help="True sequential flag for GPTQ (true or false). "
                             "Default: true",
                        default="true")
    parser.add_argument("--safetensors",
                        help="Load *.safetensors model (true or false). "
                             "If false - load *.bin model. "
                             "Currently only supported for GPTQ precisions. "
                             "Default: true",
                        default="true")
    parser.add_argument("--fused-attention",
                        help="Inject fused attention. "
                             "Currently only supported for GPTQ precisions. "
                             "Default: false",
                        default="false")
    parser.add_argument("--fused-mlp",
                        help="Inject fused MLP. "
                             "Currently only supported for GPTQ precisions. "
                             "Default: false",
                        default="false")
    parser.add_argument("--quantization-type",
                        help=f"Quantization type ({', '.join([x.value for x in QuantizationType])}). "
                             "Currently only suported for float4 precision. "
                             f"Default: {QuantizationType.NF4.value}",
                        default=QuantizationType.NF4)
    parser.add_argument("--double-quantization",
                        help="Use double quantization. "
                             "Currently only supported for float4 precision. "
                             "Default: true",
                        default="true")
    parser.add_argument("--version",
                        help="Show the version of this Neodim Server",
                        action="store_true")
    args = parser.parse_args()

    args.listen_port = int(args.listen_port)
    args.layer_strs = args.layers.split(",")
    args.precision = ModelPrecision(args.precision)
    args.group_size = int(args.group_size)
    args.true_sequential = args.true_sequential == "true"
    args.safetensors = args.safetensors == "true"
    args.fused_attention = args.fused_attention == "true"
    args.fused_mlp = args.fused_mlp == "true"
    args.quantization_type = QuantizationType(args.quantization_type)
    args.double_quantization = args.double_quantization == "true"
    return args


def main() -> None:
    args = parse_args()

    if args.version:
        print(server.SERVER_VERSION)
        return

    if not args.model:
        raise RuntimeError("--model is missing")

    print_gpu_info()
    model, tokenizer, gpu_device = load_model(args)
    run_ai_server(model, tokenizer, gpu_device, args.listen_address, args.listen_port)


if __name__ == "__main__":
    main()
